src/app/qgisapp.cpp

```diff
@@ -12811,7 +12821,17 @@ bool QgisApp::addRasterLayers( QStringList const &fileNameQStringList, bool guiW
       QFileInfo myFileInfo( *myIterator );
 
       // try to create the layer
-      QgsRasterLayer *layer = addRasterLayerPrivate( *myIterator, myFileInfo.completeBaseName(),
+      // set the layer name to the file base name...
+      QString layerName = myFileInfo.completeBaseName();
+
+      // ...unless the layer uri matches "GPKG:filePath:layerName" and layerName differs from the file base name
+      const QStringList layerUriSegments = myIterator->split( QLatin1String( ":" ) );
+      if ( layerUriSegments.count() == 3 && layerUriSegments[ 0 ] ==  QLatin1String( "GPKG" ) && layerUriSegments[ 2 ] != layerName )
```

Comment:
what about on Windows with filenames that contain ':' like c:\bla\bla

Reply:
@rouault as fas as I can see, the "GPKG:" scheme requires the layer name to be specified ("GPKG:filename" doesn't build a valid layer, even if there is a single layer in the file), so it should be safe to use the last part instead of the third one. Does it make sense?

Reply:
Nevermind, now it's replaced with QgsProviderRegistry::decodeUri()